'''
script that takes in variant matrix generated by reshape.long2wide.py, DF with variants of interest and gene annotation
aggregates variants by gene, counts number of variants (disregards zygosity)
'''

import pandas as pd
from sys import argv

script, matrix, anno, output=argv

# read in matrix with variants of interest as rows and indivduals as columns
DF=pd.read_csv(matrix, sep='\t')
POS=pd.read_csv(anno, sep='\t')

# convert chrom column to string for merging
DF.CHR=DF.CHR.astype('str')
POS.CHR=POS.CHR.astype('str')

print DF.shape
# pull out sites of interest
GENE=pd.merge(DF, POS, on=['CHR','POS','REF','ALT'])

# aggreagte genotypes by gene, count number of variants
GENE.drop(['CHR','POS','REF','ALT','REF_READS'], axis=1, inplace=True)
print GENE.head()
GENE=GENE.groupby(['GENE', 'ID']).agg(len).reset_index()

print GENE.head()
# clean up VCF IDs for eventual merging
GENE['SHORT_ID']=GENE.ID.apply(lambda x: str(x).split("-")[2])
GENE.drop('ID', axis=1, inplace=True)
GENE['TYPE']='SOM'
GENE=GENE[['SHORT_ID','GENE','ALT_READS', 'TYPE']]
GENE.columns=['SHORT_ID','GENE','GENO', 'TYPE']

GENE.to_csv(output, sep='\t',index=False)

